core/src/main/java/org/elasticsearch/plugins/PluginManager.java

```diff
@@ -260,6 +301,33 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
         }
     }
 
+    private Path unzipToTemporary(Path zip) throws IOException {
+        Path tmp = Files.createTempDirectory(environment.tmpFile(), null);
+
+        try (ZipInputStream zipInput = new ZipInputStream(Files.newInputStream(zip))) {
+            ZipEntry entry;
+            byte[] buffer = new byte[8192];
+            while ((entry = zipInput.getNextEntry()) != null) {
+                Path targetFile = tmp.resolve(entry.getName());
+
+                // be on the safe side: do not rely on that directories are always extracted
+                // before their children (although this makes sense, but is it guaranteed?)
+                Files.createDirectories(targetFile.getParent());
+                if (!entry.isDirectory()) {
+                    OutputStream out = Files.newOutputStream(targetFile);
```

Comment:
so much for using @uschindler unzip code :) 

Thanks! I will take care of these